runtime/runtime-core/common/src/aws/smithy/kotlin/runtime/net/url/UrlPath.kt

```diff
@@ -0,0 +1,209 @@
+     * affect this instance.
+     */
+    public fun toBuilder(): Builder = Builder(this)
+
+    override fun equals(other: Any?): Boolean {
+        if (this === other) return true
+        if (other == null || this::class != other::class) return false
+
+        other as UrlPath
+
+        if (segments != other.segments) return false
+        if (trailingSlash != other.trailingSlash) return false
+
+        return true
+    }
+
+    override fun hashCode(): Int {
+        var result = segments.hashCode()
+        result = 31 * result + trailingSlash.hashCode()
+        return result
+    }
+
+    override fun toString(): String = asEncoded(segments, trailingSlash)
+
+    /**
+     * A mutable builder used to construct [UrlPath] instances
+     */
+    public class Builder internal constructor(path: UrlPath?) {
+        /**
+         * Initialize an empty [UrlPath] builder
+         */
+        public constructor() : this(null)
+
+        /**
+         * Get or set the URL path as a **decoded** string.
+         */
+        public var decoded: String
+            get() = asDecoded(segments, trailingSlash)
+            set(value) { parseDecoded(value) }
+
+        /**
+         * Get or set the URL path as an **encoded** string.
+         */
+        public var encoded: String
+            get() = asEncoded(segments, trailingSlash)
+            set(value) { parseEncoded(value) }
+
+        public val segments: MutableList<Encodable> = path?.segments?.toMutableList() ?: mutableListOf()
+
+        /**
+         * A mutable list of **decoded** path segments. Any changes to this list will update the builder.
+         */
+        public val decodedSegments: MutableList<String> = segments.asView(
+            Encodable::decoded,
+            PercentEncoding.Query::encodableFromDecoded,
+        )
+
+        public fun decodedSegments(block: MutableList<String>.() -> Unit) {
+            decodedSegments.apply(block)
+        }
+
+        /**
+         * A mutable list of **encoded** path segments. Any changes to this list will update the builder.
+         */
+        public val encodedSegments: MutableList<String> = segments.asView(
+            Encodable::encoded,
+            PercentEncoding.Query::encodableFromEncoded,
+        )
+
+        public fun encodedSegments(block: MutableList<String>.() -> Unit) {
+            encodedSegments.apply(block)
+        }
+
+        /**
+         * Normalizes the segments of a URL path according to the following rules:
+         * * The returned path always begins with `/` (e.g., `a/b/c` → `/a/b/c`)
+         * * The returned path ends with `/` if the input path also does
+         * * Empty segments are discarded (e.g., `/a//b` → `/a/b`)
+         * * Segments of `.` are discarded (e.g., `/a/./b` → `/a/b`)
+         * * Segments of `..` are used to discard ancestor paths (e.g., `/a/b/../c` → `/a/c`)
+         * * All other segments are unmodified
+         */
+        public fun normalize() {
```

Comment:
The SDK does not call this anywhere by default except during signing (on a copy), Still, I added it here because it seems useful for users to be able to normalize paths which may be constructed dynamically, coming from external sources, etc.